### PR TITLE
Fix for Mac Catalyst

### PR DESCRIPTION
**Pull request by 0x0c**
Hi, I fixed the issue that this repo could not build using Mac Catalyst.
To address this issue, I added `#if targetEnvironment(macCatalyst)` to ignore using `[NSLayoutManager hyphenationFactor]` that is deprecated in iOS13+.

### CODE REVIEW

File: Examples/Example/Snapshot Tests/BaseSnapshotTestCase.swift [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) 2018-present Frédéric Maquin <fred@ephread.com> and contributors.
 // Licensed under the terms of the MIT License.
 
+#if targetEnvironment(macCatalyst)
```
Comment (ephread) [inline Examples/Example/Snapshot Tests/BaseSnapshotTestCase.swift:4]: Quick question about the snapshot tests and catalyst. Are they conditionally excluded because `iOSSnapshotTestCase` doesn't support catalyst? Did you run into issues?

File: Examples/Example/Sources/Core/Custom Views/CustomCoachMarkBodyView.swift [hunk 2]
```diff
@@ -50,7 +50,10 @@ internal class CustomCoachMarkBodyView: UIView, CoachMarkBodyView {
         self.hintLabel.font = UIFont.systemFont(ofSize: 15.0)
         self.hintLabel.isScrollEnabled = false
         self.hintLabel.textAlignment = .justified
+        #if targetEnvironment(macCatalyst)
+        #else
         self.hintLabel.layoutManager.hyphenationFactor = 1.0
```
Comment (ephread) [inline Examples/Example/Sources/Core/Custom Views/CustomCoachMarkBodyView.swift:55]: :+1: I'll do a sweep after this PR to remove them, I didn't realised it was deprecated in iOS 14 and unavailable in Catalyst.

File: Examples/InstructionsExample.entitlements [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.security.app-sandbox</key>
```
Comment (ephread) [inline Examples/InstructionsExample.entitlements:5]: Are these entitlements required by Catalyst?

**Review by ephread — COMMENTED** [review]
I never used Catalyst, and never thought about using Instructions in Catalyst, so thanks a lot for this PR! I asked a couple of questions, because I'm curious about what it entails.

I'm requesting changes only because of the `.DS_Store` that creeped in the PR. It needs to be removed. I have them ignored in a global gitconfig, but you can add it to the project gitignore if you want.

Everything else looks good to me, thanks again!

File: Examples/Example/Snapshot Tests/BaseSnapshotTestCase.swift (hunk 1 above)
Comment (0x0c) [reply Examples/Example/Snapshot Tests/BaseSnapshotTestCase.swift:4]: I could not use `iOSSnapshotTestCase` on macOS as following error.
Screen Shot 2020-10-05 at 13 59 56
I tried `carthage update --platform macOS` but I got an error `Dependency "ios-snapshot-test-case" has no shared framework schemes for any of the platforms: Mac`.

I think `iOSSnapshotTestCase` does not support macOS.

**Comment by 0x0c** [comment]
Thank you for your review!

> I'm requesting changes only because of the .DS_Store that creeped in the PR. It needs to be removed. I have them ignored in a global gitconfig, but you can add it to the project gitignore if you want.

OK, I'll add new `.gitignore` which is generated from gitignore.io.

File: Examples/InstructionsExample.entitlements (hunk 3 above)
Comment (0x0c) [reply Examples/InstructionsExample.entitlements:5]: To run an app on macOS Catalina, we have to notarize the app. This file is added by Xcode when I ran an example app on my mac. So I think, as you mentioned, the entitlements required by Catalyst.

File: Examples/Example/Sources/Core/Custom Views/CustomCoachMarkBodyView.swift (hunk 2 above)
Comment (0x0c) [reply Examples/Example/Sources/Core/Custom Views/CustomCoachMarkBodyView.swift:55]: I got it!

**Comment by 0x0c** [comment]
I can't understand why GitHub action fail :(
This commit (bba190c) passed the action and I didn't change any code.

**Comment by ephread** [comment]
@0x0c I'll have a look, sometimes it's just because XCUITest isn't great great and there's a race condition I've never been able to eliminate. It won't prevent the merge.

That said, 👍 for using a generated .gitignore, but in the process you removed some of the ignores I had put there previously (the most important one are `**/Carthage/Checkouts` and `**/Carthage/Build`. Can you add re-add these ignores in the right section? Thanks!

```
## Build generated
Tests

## Other
*.xccheckout
*.moved-aside
*.xcuserstate
*.xcscmblueprint

## Carthage
**/Carthage/Checkouts
**/Carthage/Build

Skassets
.swiftpm/
```

Hang in there, it's nearly merged. Thanks for your work!

**Comment by 0x0c** [comment]
@ephread Thanks for your quick reply!
OK, I'll add some lines to `.gitignore` as you mentioned, but I have some questions and need your opinions before adding the lines you suggested.

First, there are some lines to ignore files that are generated by Carthage(See line 9 and line 11 in `.gitignore`). Isn't it enough? (Don't worry, I'll remove `#` from line 9.)
Second, what is purpose of adding `xcuserstate`. I think `*.xcuserstate` files may be ignored by `xcuserdata/` at line 69. (See an attached image.)
Third, what is purpose of adding `Tests` and `Skassets`? I've never seen that and I'm interested about advantages for ignoring these files.

Thank you!

FYI: These lines are already added by gitignore.io :)
```
*.xccheckout
*.moved-aside
*.xcscmblueprint
```
See https://github.com/0x0c/Instructions/blob/main/.gitignore#L72 to https://github.com/0x0c/Instructions/blob/main/.gitignore#L78.

Screen Shot 2020-10-15 at 2 07 28

**Comment by ephread** [comment]
@0x0c 
> Thanks for your quick reply!

Cheers!

> I have some questions and need your opinions before adding the lines you suggested.

Sure, happy to answer them. 🙂

> First, there are some lines to ignore files that are generated by Carthage(See line 9 and line 11 in `.gitignore`). Isn't it enough? (Don't worry, I'll remove # from line 9.)

It might just be. The original reason for the double stars is that `Carthage` is nested somewhere inside `Example/`. I don't remember whether I encountered some issues, if this used to be a requirement in an older version of git or if it's actually the appropriate way to target nested directories.

> Second, what is purpose of adding xcuserstate? I think *.xcuserstate files may be ignored by xcuserdata/ at line 69. (See an attached image.)

Yes, I think you're right. It's probably redundant. I simply didn't want to have stuff removed from the `.gitgnore` and accidentally commit files that were previously ignored. I'm being busy-lazy, since it's working, I don't want to break it.

> Third, what is purpose of adding Tests and Skassets? I've never seen that and I'm interested about advantages for ignoring these files.

So these are tailored to Instructions. `Tests` is generated by [Slather](https://github.com/ephread/Instructions/blob/9aeb31c4ed4202f9657ffcd86b53edd475452516/.slather.yml) and `SKAssets` is just and old directory I have on my local setup with Sketch resources.

> FYI: These lines are already added by gitignore.io :)

Sweet, I missed them, that's a good catch!

Bottom line, the less change, the better!

**Comment by 0x0c** [comment]
Oh, I find the reason of first question. The files that are generated by Carthage are not placed at the top level of this repository and it isn't enough to ignore the files. So, I suggest adding new `.gitignore` to inside of `Example` directory, is it ok?

**Comment by 0x0c** [comment]
Oh no, we got a race condition. I post my comment after you replied before I reload this page. Sorry.
I got it and I'll add the lines at the end of `.gitignore`.

How about my suggestion? 
> The files that are generated by Carthage are not placed at the top level of this repository and it isn't enough to ignore the files. So, I suggest adding new .gitignore to inside of Example directory, is it ok? It may same file as current .gitignore.

**Comment by 0x0c** [comment]
@ephread Oh no, we got a race condition. I post my comment after you replied before I reload this page. Sorry.
I got it and I'll add the lines at the end of `.gitignore`.

How about my suggestion? 
> The files that are generated by Carthage are not placed at the top level of this repository and it isn't enough to ignore the files. So, I suggest adding new .gitignore to inside of Example directory, is it ok? It may same file as current .gitignore.

**Comment by ephread** [comment]
@0x0c Haha, no worries 😂 We're like XCUITest now!

> How about my suggestion?

I think splitting the .gitignore will make things harder to maintain for me in the long run so I would prefer to keep everything in a top level file. I think `**/Carthage` would also match a top level `Carthage` directory, but feel free to have both patterns if it makes more sense.

**Comment by 0x0c** [comment]
@ephread I agree your opinion. I updated `.gitignore` just now. Please take a look :) https://github.com/0x0c/Instructions/commit/e18915bf9b23a5b377209d31b90db47cc8945a72

I thank for your polite reviews!

**Comment by ephread** [comment]
@0x0c almost there! The SwiftPM and Carthage files previous ignored are still committed in the PR, just need to remove them and we're good to go!

> I thank for your polite reviews!

All good, thanks for working on this!

**Comment by 0x0c** [comment]
@ephread I removed `Example/Carthage/` and `.swiftpm` directory. I made mistakes removing files and I noticed after push to `main`. So I checked out 4065c10 and make `HEAD` as `main` to remove some commits, and force-push to my repository.

**Comment by ephread** [comment]
Looks good to me. I'm going to merge, have a look at the tests later this week and hopefully release a new version this weekend! 🎉

Thanks again for your great work!

**Comment by 0x0c** [comment]
Thanks a lot!

**Comment by DisobedientMedia** [comment]
Hi ephread. Thanks for looking into merging this. What's the current status? I'm hitting this problem myself today.

**Comment by ephread** [comment]
> This change isnt coming up in my Swift Package Manager version. Do you need to make a release for this. Am I missing something - only just moved to SWP and still working it out.

@DisobedientMedia you probably need to point to the `main` branch in Xcode / Swift Package Manager as it's not released yet. I planned to create a new release a while ago, but it got postponed because I wanted to fix the race condition found in the tests first. I'll try to make a release this weekend regardless!
